pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/ParametricRuleViolation.java

```diff
@@ -183,11 +192,31 @@ public String getVariableName() {
         return variableName;
     }
 
+    @Experimental
+    @Override
+    public String getSourceCode(String encoding) {
+        try {
+            SourceCode sourceCode = new SourceCode(new SourceCode.StringCodeLoader(rawCodeSnippet));
```

Comment:
... The only reasonable way I see to implement this in pmd 6 is to use a FileCodeLoader here and load the filename, then slice the contents of the file. This should not be done eagerly as it sounds really bad for performance.

Reply:
I used StringCodeLoader here purely for testing purposes used by XMLVerboseRendererTest. Using FileCodeLoader caused an exception during testing.

Reply:
Yes but then your rawCodeSnippet is not the code snippet, it's the entire source file, since you slice it late. We can't keep all analysed files in memory just for this renderer to work...

Reply:
I intended my changes in ParametricRuleViolation class to be only used in unit testing. I can create a separate class for unit testing instead.